### PR TITLE
[HOTFIX] Fix INSERT STAGE footer read error

### PR DESCRIPTION
**Pull request by jackylk**
### Why is this PR needed?
 There are 2 issues:
1. INSERT STAGE will skip bytes to read the last long in the carbondata file as the offset to the footer, but sometimes DFS is not skipping the right byte size
2. When error occurs, segment status is not in SUCCESS state, but recovery handling is not doing correctly in  INSERT STAGE command
 
 ### What changes were proposed in this PR?
1. In CarbonFooterReaderV3.readFooterVersion3, if offset is not set by caller, read it from corresponding index file.
2. In CarbonInsertFromStageCommand, when recovery is required, delete the segment entry if the entry is not in SUCESS state.
    
 ### Does this PR introduce any user interface change?
No

 ### Is any new testcase added?
No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1459/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1461/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1462/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1470/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1471/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1479/

**Comment by niuge01** [comment]
LGTM

File: core/src/main/java/org/apache/carbondata/core/reader/CarbonFooterReaderV3.java [hunk 1]
```diff
@@ -59,17 +55,20 @@ public FileFooter3 readFooterVersion3() throws IOException {
     thriftReader.open();
 
     // If footer offset is 0, means caller does not set it,
-    // so we read it from the end of the file
+    // so we read it from the index file
     if (footerOffset == 0) {
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/reader/CarbonFooterReaderV3.java]: The function of this method should be single, it should only support reading footer.

Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/reader/CarbonFooterReaderV3.java]: if it needs to read the index, better to implement in another place where it decides to read index or footer.

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1526/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1527/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1530/

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 2]
```diff
@@ -714,4 +714,16 @@ public static String generateBadRecordsPath(String badLogStoreLocation, String s
           CarbonCommonConstants.FILE_SEPARATOR + taskNo;
     }
   }
+
+  /**
+   * Return the parent path of the input file
+   */
+  public static String getParentPath(String dataFilePath) {
+    int endIndex = dataFilePath.lastIndexOf(CarbonCommonConstants.FILE_SEPARATOR);
```
Comment (qiuchenjian) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: CarbonCommonConstants.FILE_SEPARATOR is the last character of dataFilePath, such as '/user/warehouse/carbon.store/www/'

File: core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java [hunk 3]
```diff
@@ -112,4 +112,34 @@ public void openThriftReader(byte[] fileData) throws IOException {
...
+  /**
+   * Return the footer offset of the given dataFileName in the index file
+   *
+   * @param indexFilePath path of the index file
+   * @param dataFileName file name of the data file to match in the index file
+   * @param conf hadoop configuration
+   * @return footer offset of the data file
+   * @throws IOException
+   */
+  public static long readFooterOffsetFromIndexFile(String indexFilePath, String dataFileName,
```
Comment (kunal642) [inline core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java]: This method is unused..Please remove if not needed

Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java (hunk 2 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: the input is a file path like `/user/warehouse/carbon.store/www/file.carbondata`, so this function will return `/user/warehouse/carbon.store/www/`

Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: I updated the comment in this function

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1551/

File: core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java [hunk 4]
```diff
@@ -112,4 +112,5 @@ public void openThriftReader(byte[] fileData) throws IOException {
   public boolean hasNext() throws IOException {
     return thriftReader.hasNext();
   }
+
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java]: please revert this change

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 5]
```diff
@@ -714,4 +714,18 @@ public static String generateBadRecordsPath(String badLogStoreLocation, String s
...
   }
+
+  /**
+   * Return the parent path of the input file.
+   * For example, if input file path is /user/warehouse/t1/file.carbondata
+   * then return will be /user/warehouse/t1
+   */
+  public static String getParentPath(String dataFilePath) {
+    int endIndex = dataFilePath.lastIndexOf(CarbonCommonConstants.FILE_SEPARATOR);
+    if (endIndex > -1) {
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: can we replace if else with conditional operator here?

File: core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java (hunk 4 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/reader/CarbonIndexFileReader.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java (hunk 5 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: I think this is better, especially for debugging

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1571/

**Comment by QiangCai** [comment]
LGTM

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1650/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1670/

**Comment by kunal642** [comment]
LGTM
